File: server/app/schemas/ticket.py
"""
Pydantic schemas for ticket CRUD + workspace queries
"""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.base import TimestampSchema, PaginatedResponse
from app.schemas.ticket_view import ConditionItem

CustomFieldValue = str | int | float | bool | list | dict | None


class RelatedTicketResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ticket_number: str | None = None


class TicketResponse(TimestampSchema):
    model_config = ConfigDict(from_attributes=True)
    id: int
    tenant_id: int
    ticket_number: str | None = None
    layout_id: int | None = None
    conversation_id: int | None = None
    conversation_public_id: str | None = None
    call_record_id: int | None = None
    call_record_call_id: str | None = None
    user_id: int | None = None
    agent_id: int | None = None
    assignee_group_id: int | None = None
    title: str
    description: str | None = None
    status: str = "open"
    priority: str | None = None
    created_by: dict | None = None
    updated_by: dict | None = None
    custom_fields: dict[str, CustomFieldValue] = Field(default_factory=dict)


class TicketCreate(BaseModel):
    """Create a new ticket — system + custom fields."""
    title: str = Field(..., min_length=1, max_length=256)
    description: str | None = Field(None, max_length=5000)
    status: str = Field(default="open", max_length=32)
    priority: str | None = Field(default="medium", max_length=16)
    layout_id: int | None = None
    conversation_id: int | None = None
    call_record_id: int | None = None
    user_id: int | None = None
    agent_id: int | None = None
    assignee_group_id: int | None = None
    custom_fields: dict[str, CustomFieldValue] = Field(default_factory=dict)


class TicketUpdate(BaseModel):
    """Partial update of a ticket."""
    title: str | None = Field(None, min_length=1, max_length=256)
    description: str | None = Field(None, max_length=5000)
    status: str | None = Field(None, max_length=32)
    priority: str | None = Field(None, max_length=16)
    conversation_id: int | None = None
    call_record_id: int | None = None
    user_id: int | None = None
    agent_id: int | None = None
    assignee_group_id: int | None = None
    custom_fields: dict[str, CustomFieldValue] | None = None


class TicketListResponse(PaginatedResponse):
    items: list[TicketResponse]


class TicketQueryRequest(BaseModel):
    """POST body for querying tickets with optional view + temp filters."""
    view_id: int | None = None
    search: str | None = Field(None, max_length=256)
    temp_conditions: list[ConditionItem] = Field(default_factory=list)
    temp_condition_logic: str = Field(default="and", max_length=8)
    group_value: str | None = None
    sort_by: str | None = None
    sort_order: str = Field(default="desc", max_length=4)
    page: int = 1
    per_page: int = 20


class TicketViewCountItem(BaseModel):
    view_id: int
    count: int


class TicketViewCountsResponse(BaseModel):
    total_count: int = 0
    items: list[TicketViewCountItem]

File: web/models/ticket.ts
import type { ConditionItem } from '@/models/ticket-view'
import type { CustomFieldValue } from '@/types/custom-field-value'

export type { CustomFieldValue } from '@/types/custom-field-value'

export type Ticket = {
  id: number
  tenant_id: number
  ticket_number: string | null
  layout_id: number | null
  conversation_id: number | null
  conversation_public_id: string | null
  call_record_id: number | null
  call_record_call_id: string | null
  user_id: number | null
  agent_id: number | null
  assignee_group_id: number | null
  title: string
  description: string | null
  status: string
  priority: string | null
  created_by: AuditActorRef | null
  updated_by: AuditActorRef | null
  custom_fields: Record<string, CustomFieldValue>
  created_at: string
  updated_at: string
}

export type RelatedTicket = {
  id: number
  ticket_number: string | null
}

export type AuditActorRef = {
  actor_type: string | null
  actor_id: number | null
  actor_name: string | null
}

export type TicketChangeValue =
  | CustomFieldValue
  | Record<string, unknown>
  | unknown[]

export type TicketChangeEntry = {
  field_key: string
  field_label: string
  old_value: TicketChangeValue
  new_value: TicketChangeValue
}

export type TicketChange = {
  id: number
  tenant_id: number
  ticket_id: number
  actor_type: string
  actor_id: number | null
  actor_name: string | null
  /** Present when the actor is an employee with an uploaded profile photo. */
  actor_avatar?: string | null
  field_key: string
  field_label: string
  field_source: string
  old_value: TicketChangeValue
  new_value: TicketChangeValue
  /** Batched field updates from a single save (field_key = __batch__) */
  entries?: TicketChangeEntry[] | null
  created_at: string
  updated_at: string
}

export type CreateTicketPayload = {
  title: string
  description?: string | null
  status?: string
  priority?: string | null
  layout_id?: number | null
  conversation_id?: number | null
  call_record_id?: number | null
  user_id?: number | null
  agent_id?: number | null
  assignee_group_id?: number | null
  custom_fields?: Record<string, CustomFieldValue>
}

export type UpdateTicketPayload = {
  title?: string | null
  description?: string | null
  status?: string | null
  priority?: string | null
  conversation_id?: number | null
  call_record_id?: number | null
  user_id?: number | null
  agent_id?: number | null
  assignee_group_id?: number | null
  custom_fields?: Record<string, CustomFieldValue>
}

export type TicketQueryPayload = {
  view_id?: number | null
  search?: string | null
  temp_conditions?: ConditionItem[]
  temp_condition_logic?: 'and' | 'or'
  group_value?: string | null
  sort_by?: string | null
  sort_order?: 'asc' | 'desc'
  page?: number
  per_page?: number
}

export type TicketViewCountItem = {
  view_id: number
  count: number
}

export type TicketViewCountsResponse = {
  total_count: number
  items: TicketViewCountItem[]
}
